nvbench/axis_iteration_space.cuh

```diff
@@ -0,0 +1,93 @@
+/*
+ *  Copyright 2021 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 with the LLVM exception
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.
+ *
+ *  You may obtain a copy of the License at
+ *
+ *      http://llvm.org/foundation/relicensing/LICENSE.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#pragma once
+
+#include <nvbench/detail/axes_iterator.cuh>
+
+namespace nvbench
+{
+
+struct axis_space_base
+{
+  using axes_type = std::vector<std::unique_ptr<nvbench::axis_base>>;
+  using axes_info = std::vector<detail::axis_index>;
+
+  using AdvanceSignature =
+    nvbench::detail::axis_space_iterator::AdvanceSignature;
+  using UpdateSignature = nvbench::detail::axis_space_iterator::UpdateSignature;
+
+  axis_space_base(std::vector<std::size_t> input_indices,
+                  std::vector<std::size_t> output_indices);
+  virtual ~axis_space_base();
+
+  [[nodiscard]] std::unique_ptr<axis_space_base> clone() const;
+  [[nodiscard]] std::vector<std::unique_ptr<axis_space_base>>
+  clone_as_linear() const;
+
+  [[nodiscard]] detail::axis_space_iterator iter(const axes_type &axes) const;
+  [[nodiscard]] std::size_t size(const axes_type &axes) const;
+  [[nodiscard]] std::size_t valid_count(const axes_type &axes) const;
+
+  [[nodiscard]] bool contains(std::size_t input_index) const;
+
+protected:
+  std::vector<std::size_t> m_input_indices;
+  std::vector<std::size_t> m_output_indices;
+
+  virtual std::unique_ptr<axis_space_base> do_clone() const         = 0;
+  virtual detail::axis_space_iterator do_iter(axes_info info) const = 0;
+  virtual std::size_t do_size(const axes_info &info) const          = 0;
+  virtual std::size_t do_valid_count(const axes_info &info) const   = 0;
```

Comment:
It'd be nice to get some docs on the concept of a "axis space" as well as the semantics of these functions.

Reply:
I have added more documentation